embassy-stm32/Cargo.toml

```diff
@@ -33,7 +33,7 @@ flavors = [
 
 [dependencies]
 embassy-sync = { version = "0.3.0", path = "../embassy-sync" }
-embassy-time = { version = "0.1.3", path = "../embassy-time", optional = true }
+embassy-time = { version = "0.1.3", path = "../embassy-time" }
```

Comment:
Won't this require everyone to enable a time driver in order to use the stm32 HAL? 🤔 I don't think that should be enforced.

Reply:
Alternatively I'd suggest to add a `dma` feature that requires `time` in order to implement timeout support. Not having timeout functionality is too big of a footgun IMO.

Reply:
Otherwise I'm open to suggestions to implement timeouts with a different approach.

Reply:
I've not spent much time with stm32 i2c, but is it not possible to handle timeouts externally, like this?

```
match select(Timer::after(Duration::from_secs(10)), i2c.write(...)).await {
    Either::First(_) => {} // Timeout!
    Either::Second(_) => {} // Success
}
```

Iirc, dropping the write future should do the required cancellation

Reply:
Thanks, I think I understand your arguments better now. So, I think these are potential ways forward then:

* Guard I2C fns using timeouts behind the 'time' feature and make embassy-time optional again. This way you _have_ to enable timekeeping to use i2c. I think it's fine if there is never a case where you don't need timeouts. 
* If there is a valid use case to not use timeouts, then I think the better approach is to provide 2 methods: foo() and foo_with_timeout() where the second one is guarded by the time feature.

I don't know the entire set of i2c use cases, so I don't know which one of these options are best. If there are no reservations from others against the first option, I would say go for that. Most important I think is to allow the use case of not enabling embassy-time, even if that means no i2c.

Reply:
I think guarding the APIs behind a feature guard is a reasonable fix. The question the would be, do we want them to be guarded by a `time` feature or by an `async-i2c` (or something else) feature that requires `time`?

Reply:
Good question... I personally think using the `time` feature is fine, but probably warrants a comment in the code?